src/oracle/oracle.cairo

```diff
@@ -379,22 +771,53 @@ mod Oracle {
             price: u128,
             ref_price: u128,
             max_ref_price_deviation_factor: u128,
-        ) { // TODO
+        ) {
+            let diff = calc::diff(price, ref_price);
+
+            let diff_factor = precision::to_factor(diff, ref_price);
+
+            if diff_factor > max_ref_price_deviation_factor {
+                OracleError::MAX_REFPRICE_DEVIATION_EXCEEDED(
+                    token, price, ref_price, max_ref_price_deviation_factor
+                );
+            }
         }
 
         /// Set the primary price.
         /// # Arguments
         /// * `token` - The token to set the price for.
         /// * `price` - The price value to set to.
-        fn set_primary_price(
-            ref self: ContractState, token: ContractAddress, price: Price,
-        ) { // TODO
+        fn set_primary_price_(ref self: ContractState, token: ContractAddress, price: Price) {
+            self.primary_prices.write(token, price);
         }
 
         /// Remove the primary price.
         /// # Arguments
         /// * `token` - The token to set the price for.
-        fn remove_primary_price(ref self: ContractState, token: ContractAddress) { // TODO
+        fn remove_primary_price(ref self: ContractState, token: ContractAddress) {
```

Comment:
I think we can find something more optimized. Maybe getting inspiration from what's being done in those PRs:
https://github.com/keep-starknet-strange/satoru/pull/312
https://github.com/keep-starknet-strange/satoru/pull/295

Reply:
You mean using a`LagacyMap::<u32, ContractAddress>` instead of a `List<ContractAddress>` for `token_with_prices`?

Reply:
List has been kept but the internals of the fn have been refactored.

Reply:
changed to LagacyMap::<u32, ContractAddress>